File: backend/app/core/estimation/destatis_index.py
"""
§50 — Indice Destatis officiel (61261-0002, bpr110), base 2021 = 100, inkl. USt.
Publication du 10.07.2026.

Miroir backend de la même série que le frontend (src/data/destatisIndex.ts) —
UNE série par couche, épinglée par des tests des DEUX côtés avec les mêmes
valeurs officielles. Sert à indexer les prix du bureau (Preisstand JJJJ) vers
l'année courante : méthode standard des ingénieurs coûts (Preisspiegelung).
"""

from __future__ import annotations

from decimal import Decimal, ROUND_HALF_UP
from typing import Dict, List, Optional, Tuple

# (année, trimestre) → indice officiel, 1 décimale. Stand 10.07.2026.
_QUARTERLY: List[Tuple[int, int, Decimal]] = [
    (2021, 1, Decimal("96.7")), (2021, 2, Decimal("99.5")),
    (2021, 3, Decimal("101.6")), (2021, 4, Decimal("102.2")),
    (2022, 1, Decimal("108.2")), (2022, 2, Decimal("112.1")),
    (2022, 3, Decimal("112.7")), (2022, 4, Decimal("113.1")),
    (2023, 1, Decimal("114.7")), (2023, 2, Decimal("115.7")),
    (2023, 3, Decimal("115.9")), (2023, 4, Decimal("116.2")),
    (2024, 1, Decimal("128.5")), (2024, 2, Decimal("129.4")),
    (2024, 3, Decimal("130.3")), (2024, 4, Decimal("130.8")),
    (2025, 1, Decimal("132.6")), (2025, 2, Decimal("133.6")),
    (2025, 3, Decimal("134.3")), (2025, 4, Decimal("135.0")),
    (2026, 1, Decimal("137.0")), (2026, 2, Decimal("140.3")),
]

INDEX_STAND = "10.07.2026"
INDEX_SOURCE = "Destatis 61261-0002 (bpr110), Basis 2021=100, inkl. USt"

# Facteur de rétrocétion base 2021 → base 2020 (Wechsel Genesis,
# Jahresdurchschnitt 2021 base 2020 = 112,7).
_RETRO_2021_TO_2020 = Decimal("1.127")

_CENT = Decimal("0.01")


def _q2(value: Decimal) -> Decimal:
    return value.quantize(_CENT, rounding=ROUND_HALF_UP)


def index_for_year(year: int) -> Optional[Decimal]:
    """Indice annuel : moyenne des 4 trimestres si complets, dernier point
    publié pour l'année en cours, 2020 rétrocédé ; None avant 2020."""
    if year < 2020:
        return None
    if year == 2020:
        return _q2(Decimal("100") / _RETRO_2021_TO_2020)  # 88,73
    points = [v for (y, _q, v) in _QUARTERLY if y == year]
    if len(points) >= 4:
        return _q2(sum(points) / Decimal(len(points)))
    if points:
        return points[-1]
    return None


def yearly_index() -> Dict[int, Decimal]:
    """Table annuelle 2020..(dernière année couverte)."""
    out: Dict[int, Decimal] = {}
    last_year = max(y for (y, _q, _v) in _QUARTERLY)
    for year in range(2020, last_year + 1):
        value = index_for_year(year)
        if value is not None:
            out[year] = value
    return out


def year_factor_for(base_year: int, target_year: int) -> Decimal:
    """Facteur multiplicatif base → cible, plafonné ±40 % (au-delà le
    référentiel est trop vieux : l'estimation doit le dire, pas extrapoler).
    Retourne 1 si une année n'est pas supportée."""
    base = index_for_year(base_year)
    target = index_for_year(target_year)
    if base is None or target is None or base <= 0:
        return Decimal("1")
    factor = target / base
    return max(Decimal("0.6"), min(Decimal("1.4"), factor))

File: frontend/src/data/destatisIndex.ts
/**
 * §49 Indice Destatis officiel des prix de construction — Bauliche Anlagen
 * an Wohngebäuden (Genesis 61261-0002, série bpr110), **base 2021 = 100,
 * Y COMPRIS TVA (USt)**, publication du 10.07.2026.
 *
 * Source : Statistisches Bundesamt (Destatis), « Preisindizes für die
 * Bauwirtschaft, 2. Vierteljahr 2026 » (communiqué n° 265 du 10.07.2026) —
 * https://www.destatis.de/DE/Themen/Wirtschaft/Konjunkturindikatoren/Preise/bpr110.html
 *
 * C'est la **seule série d'indices** de l'application : countries.ts
 * (COST_INDEX, facteur d'année de l'estimation rapide) et le script d'import
 * backend la consomment. Ne plus dupliquer de constantes d'indice ailleurs.
 */

export interface IndexPoint {
  /** ex. "2024-Q1" */
  quarter: string;
  /** indice officiel, base 2021 = 100, inkl. USt, 1 décimale */
  value: number;
}

/** Série trimestrielle officielle (Stand 10.07.2026). */
export const DESTATIS_WOHNGEBAEUDE: IndexPoint[] = [
  { quarter: "2021-Q1", value: 96.7 },
  { quarter: "2021-Q2", value: 99.5 },
  { quarter: "2021-Q3", value: 101.6 },
  { quarter: "2021-Q4", value: 102.2 },
  { quarter: "2022-Q1", value: 108.2 },
  { quarter: "2022-Q2", value: 112.1 },
  { quarter: "2022-Q3", value: 112.7 },
  { quarter: "2022-Q4", value: 113.1 },
  { quarter: "2023-Q1", value: 114.7 },
  { quarter: "2023-Q2", value: 115.7 },
  { quarter: "2023-Q3", value: 115.9 },
  { quarter: "2023-Q4", value: 116.2 },
  { quarter: "2024-Q1", value: 128.5 },
  { quarter: "2024-Q2", value: 129.4 },
  { quarter: "2024-Q3", value: 130.3 },
  { quarter: "2024-Q4", value: 130.8 },
  { quarter: "2025-Q1", value: 132.6 },
  { quarter: "2025-Q2", value: 133.6 },
  { quarter: "2025-Q3", value: 134.3 },
  { quarter: "2025-Q4", value: 135.0 },
  { quarter: "2026-Q1", value: 137.0 },
  { quarter: "2026-Q2", value: 140.3 },
];

function round2(v: number): number {
  return Math.round(v * 100) / 100;
}

/** Dernier point publié (dérivé de la série — ne jamais coder en dur). */
export const INDEX_LATEST = DESTATIS_WOHNGEBAEUDE[DESTATIS_WOHNGEBAEUDE.length - 1];

export const INDEX_SOURCE_LABEL = "Destatis 61261-0002 \u00b7 Basis 2021=100 \u00b7 inkl. USt";

/** « Destatis \u00b7 Stand 06/2026 » : mois de r\u00e9f\u00e9rence du trimestre (Q2 = juin).
 *  La date de publication (10.07.2026) figure dans le commentaire de source. */
export function indexStandLabel(): string {
  const [year, q] = INDEX_LATEST.quarter.split("-Q");
  const month = (Number(q) * 3).toString().padStart(2, "0");
  return `Destatis \u00b7 Stand ${month}/${year}`;
}

function yearlyFromSeries(year: number): number | null {
  const pts = DESTATIS_WOHNGEBAEUDE.filter((p) => p.quarter.startsWith(`${year}-`));
  if (pts.length >= 4) return round2(pts.reduce((s, p) => s + p.value, 0) / pts.length);
  // année incomplète (année en cours) : dernier point publié
  if (pts.length > 0) return pts[pts.length - 1].value;
  return null;
}

/** Facteur de rétrocétion base 2021 → base 2020 (Wechsel Genesis, Jahres-
 * durchschnitt 2021 base 2020 = 112,7). Publié et traçable. */
const RETRO_2021_TO_2020 = 1.127;

/**
 * Indice annuel pour l'année `year` (par ex. indexForYear(2024) = 129,75).
 * - années couvertes par 4 trimestres : moyenne annuelle officielle ;
 * - année en cours (2026) : dernier point publié (Q2/2026 = 140,3) ;
 * - 2020 : rétrocédé de la moyenne 2021 (100,0 / 1,127 = 88,73) ;
 * - avant 2020 : non supporté (null) — le référentiel serait trop vieux.
 */
export function indexForYear(year: number): number | null {
  if (year < 2020) return null;
  if (year === 2020) return round2(100 / RETRO_2021_TO_2020);
  return yearlyFromSeries(year);
}

/** Table annuelle prête à l'emploi (clé = année, 2020..2026). */
export const YEARLY_INDEX: Readonly<Record<number, number>> = (() => {
  const out: Record<number, number> = {};
  for (let y = 2020; y <= 2026; y++) {
    const v = indexForYear(y);
    if (v !== null) out[y] = v;
  }
  return out;
})();

/**
 * Facteur multiplicatif pour convertir des prix de l'année `base` (par ex. les
 * référentiels catalogués 2024) vers l'année `year`. Plafonné à ±40 % — au-delà
 * le référentiel est trop vieux et l'estimation doit le dire (uncertaintyPct).
 */
export function yearFactorFor(base: number, year: number): number {
  const from = indexForYear(base);
  const to = indexForYear(year);
  if (from === null || to === null || from <= 0) return 1;
  const f = to / from;
  return Math.min(1.4, Math.max(0.6, f));
}
